shotover/src/codec/cassandra.rs

```diff
@@ -67,96 +153,246 @@ impl CassandraDecoder {
+                    return Err(CheckFrameSizeError::NotEnoughBytes);
+                }
+
+                let body_len = i32::from_be_bytes(src[5..9].try_into().unwrap()) as usize;
+
+                let envelope_len = ENVELOPE_HEADER_LEN + body_len;
+                if src.len() < envelope_len {
+                    return Err(CheckFrameSizeError::NotEnoughBytes);
+                }
+
+                let _ = Version::try_from(src[0])
+                    .map_err(|_| CheckFrameSizeError::UnsupportedVersion(src[0] & 0x7f))?;
+
+                Ok(envelope_len)
+            }
+        }
+    }
 }
 
-fn set_compression(compression_state: &mut Arc<RwLock<Compression>>, startup: &BodyReqStartup) {
+fn set_startup_state(
+    compression_state: &mut Arc<AtomicCompressionState>,
+    version_state: &mut Arc<AtomicVersionState>,
+    version: Version,
+    startup: &BodyReqStartup,
+) {
     if let Some(compression) = startup.map.get("COMPRESSION") {
-        let mut write = compression_state.write().unwrap();
-
-        *write = match compression.as_str() {
-            "snappy" | "SNAPPY" => Compression::Snappy,
-            "lz4" | "LZ4" => Compression::Lz4,
-            "" | "none" | "NONE" => Compression::None,
-            _ => panic!(),
-        };
+        compression_state.store(
+            match compression.as_str() {
+                "snappy" | "SNAPPY" => Compression::Snappy,
+                "lz4" | "LZ4" => Compression::Lz4,
+                "" | "none" | "NONE" => Compression::None,
+                _ => unimplemented!(),
+            }
+            .into(),
+            Ordering::Relaxed,
+        );
     }
+
+    version_state.store(version.into(), Ordering::Relaxed);
 }
 
 impl Decoder for CassandraDecoder {
     type Item = Messages;
     type Error = CodecReadError;
 
     fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, CodecReadError> {
+        let version: Version = self.version.load(Ordering::Relaxed).into();
+        let compression: Compression = self.compression.load(Ordering::Relaxed).into();
+        let handshake_complete = self.handshake_complete.load(Ordering::Relaxed);
+
         loop {
-            match RawCassandraFrame::check_envelope_size(src) {
+            match self.check_size(src, version, compression, handshake_complete) {
                 Ok(frame_len) => {
-                    // Clear the read bytes from the FramedReader
-                    let bytes = src.split_to(frame_len);
-                    tracing::debug!(
-                        "{}: incoming cassandra message:\n{}",
-                        self.direction,
-                        pretty_hex::pretty_hex(&bytes)
-                    );
-
-                    let version = Version::try_from(bytes[0])
-                        .expect("Gauranteed because check_envelope_size only returns Ok if the Version will parse");
-                    if let Version::V3 | Version::V4 = version {
```

Comment:
Because new versions can be added without a breaking bump to `cassandra-protocol`s we should keep these:
```rust
if let Version::V3 | Version::V4 | Version::V5 = version {
```

Also we probably want to make the Version::V5 dependent on `alpha-transforms` unless we want to land benchmarks first.

Reply:
We can't know what the version byte will be in new versions though